src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/ChecksumUtilities.cs

```diff
@@ -17,15 +19,32 @@ public static string BytesToString(ImmutableArray<byte> bytes)
             throw new ArgumentNullException(nameof(bytes));
         }
 
-        using var _ = StringBuilderPool.GetPooledObject(out var builder);
-        builder.EnsureCapacity(bytes.Length);
+#if NET9_0_OR_GREATER
+        var bytesArray = ImmutableCollectionsMarshal.AsArray(bytes)!;
+
+        return Convert.ToHexStringLower(bytesArray);
+#else
+        const int StackAllocThreshold = 256; // reasonable for stackalloc
+        var charCount = bytes.Length * 2;
 
+        // As this should be getting called with a Checksum array of length 32, this shouldn't allocate
+        var buffer = charCount <= StackAllocThreshold
+            ? stackalloc char[charCount]
+            : new char[charCount];
+
+        var bufferIndex = 0;
         foreach (var b in bytes)
         {
-            // The x2 format means lowercase hex, where each byte is a 2-character string.
-            builder.Append(b.ToString("x2", CultureInfo.InvariantCulture));
+            // Write hex chars directly
+            buffer[bufferIndex++] = GetHexChar(b >> 4);
+            buffer[bufferIndex++] = GetHexChar(b & 0xF);
         }
 
-        return builder.ToString();
+        // Allocate the final string
+        return buffer.ToString();
```

Comment:
Since the character count is known, consider using `string.Create(...)` to avoid copying the characters twice -- once to the stack and then again to the string's char array.

Reply:
Yep, your way is better. Switched.